### PR TITLE
Test syslinux against Debian, Fedora, Qubes configs

### PR DESCRIPTION
**Pull request by hugelgupf**
Tests parsing actual Debian, Qubes, and Fedora syslinux configs as pulled from installer ISOs and disk images.

### CODE REVIEW

**Comment by hugelgupf** [comment]
This is a PR you'll want to review commit-by-commit rather than looking at the whole thing.

**Comment by hugelgupf** [comment]
I pulled out #1699 and #1700 as smaller reviews. Will rebase when merged.

File: pkg/boot/boottest/json.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
+// Copyright 2018 the u-root Authors. All rights reserved
```
Comment (rjoleary) [inline pkg/boot/boottest/json.go:1]: 2020

File: pkg/boot/boottest/same_image.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
+// Copyright 2017-2018 the u-root Authors. All rights reserved
```
Comment (rjoleary) [inline pkg/boot/boottest/same_image.go:1]: 2020

File: pkg/boot/syslinux/config_test.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
+// Copyright 2017-2018 the u-root Authors. All rights reserved
```
Comment (rjoleary) [inline pkg/boot/syslinux/config_test.go:1]: 2020

File: pkg/boot/syslinux/testdata/debian-10-install/isolinux/addrk.cfg [hunk 4]
```diff
@@ -0,0 +1,9 @@
+label expertdark
```
Comment (rjoleary) [inline pkg/boot/syslinux/testdata/debian-10-install/isolinux/addrk.cfg:1]: Can you add a link to where these scripts came from and whether they have a license?

File: pkg/boot/syslinux/syslinux.go [hunk 5]
```diff
@@ -87,12 +88,18 @@ func ParseLocalConfig(ctx context.Context, diskDir string) ([]boot.OSImage, erro
 //
 // `s` is used to fetch any files that must be parsed or provided.
 //
-// `wd` is the default scheme, host, and path for any files named as a
-// relative path - e.g. kernel, include, and initramfs paths are requested
-// relative to the wd.
-func ParseConfigFile(ctx context.Context, s curl.Schemes, url string, wd *url.URL) ([]boot.OSImage, error) {
-	p := newParser(wd, s)
-	if err := p.appendFile(ctx, url); err != nil {
+// rootdir is the partition mount point that syslinux is operating under.
```
Comment (rjoleary) [inline pkg/boot/syslinux/syslinux.go:91]: This commit about the rootdir, can it be in a separate PR?

File: pkg/boot/boottest/json.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/u-root/u-root/pkg/boot"
+	"github.com/u-root/u-root/pkg/curl"
+)
+
+func module(r io.ReaderAt) map[string]interface{} {
```
Comment (rjoleary) [inline pkg/boot/boottest/json.go:18]: Could this instead be an interface function. I seems a bit inflexible with the type checks.

File: pkg/boot/boottest/json.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+// CompareImagesToJSON in tests.
+func ToJSONFile(imgs []boot.OSImage, filename string) error {
+	enc, err := json.MarshalIndent(ImagesToJSONLike(imgs), "", "  ")
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(filename, enc, 0644)
+}
+
+// ImagesToJSONLike spits out a json-convertible reproducible representation of
```
Comment (rjoleary) [inline pkg/boot/boottest/json.go:53]: This feels like a lot of overhead. Why not call MarshalIndent directly on boot.OSImage?

File: pkg/boot/syslinux/testdata/debian-10-install/isolinux/addrk.cfg (hunk 4 above)
Comment (hugelgupf) [reply pkg/boot/syslinux/testdata/debian-10-install/isolinux/addrk.cfg:1]: I installed Debian 10 in a VM. These are the syslinux scripts used by the ISO installer. I'm not sure they have a license...

File: pkg/boot/syslinux/syslinux.go (hunk 5 above)
Comment (hugelgupf) [reply pkg/boot/syslinux/syslinux.go:91]: No, it's a fix necessary for the Debian 9 image to work. In fact, I update the JSON for the image here: https://github.com/u-root/u-root/pull/1696/commits/e62e4ef0bd9869de38e6c796abfe4d3ddb14e9cf

File: pkg/boot/boottest/json.go (hunk 7 above)
Comment (hugelgupf) [reply pkg/boot/boottest/json.go:53]: 1. ReaderAt's don't serialize into json nicely.

2. It also adds the image_type. You wouldn't be able to distinguish a linux and multiboot image if they both only had the Name, Kernel, and Cmdline fields present.

Compare:

```
2020/06/03 22:05:54 bare image in json:                   
{                                                                                                                   
  "Name": "Test this ^media \u0026 start Fedora-Workstation-Live 27",                                               
  "Kernel": {                                                                                                       
    "ReaderAt": {}                                        
  },                                                                                                                
  "Initrd": {                                                                                                       
    "ReaderAt": {}                                                                                                  
  },                                             
  "Cmdline": "initrd=initrd.img root=live:CDLABEL=Fedora-WS-Live-27-1-6 rd.live.image rd.live.check quiet"          
}                                                                                                                   
2020/06/03 22:05:54 our json:                                                                                                                                                                                                           
{                                                                                                                   
  "cmdline": "initrd=initrd.img root=live:CDLABEL=Fedora-WS-Live-27-1-6 rd.live.image rd.live.check quiet",         
  "image_type": "linux",                                                                                            
  "initrd": {                                      
    "url": "file://testdata/fedora-27-install/isolinux/initrd.img"                                                                                                                                                                      
  },                                                                                                                
  "kernel": {                                                                                                       
    "url": "file://testdata/fedora-27-install/isolinux/vmlinuz"                                                     
  },                                                                                                                
  "name": "Test this ^media \u0026 start Fedora-Workstation-Live 27"                                                
}                             
```

File: pkg/boot/boottest/json.go (hunk 6 above)
Comment (hugelgupf) [reply pkg/boot/boottest/json.go:18]: I'd rather not, (a) so it doesn't bloat code size -- this package is only used for tests, and (b) so nobody gets the idea to use this terrible format in production somewhere.

Comment (hugelgupf) [reply pkg/boot/boottest/json.go:18]: do you want me to leave a comment in the code for this?

**Comment by hugelgupf** [comment]
I updated the code itself with answers to your questions, too. PTAL

File: pkg/boot/boottest/json.go (hunk 6 above)
Comment (hugelgupf) [reply pkg/boot/boottest/json.go:30]: Left a comment in the pkg comment.
